### ISSUE TITLE
bug: OptionSet autocommand and nvim_set_hl clear the intro screen

### ISSUE THREAD

**Issue by mikesmithgh** [neovim/neovim#26677]
### Problem

The autocommand `OptionSet` with a callback setting a highlight group, clears the intro screen on startup.

I tested on `NVIM v0.9.4` and cannot reproduce the issue.

Originally reported to https://github.com/nvim-tree/nvim-web-devicons/issues/355, then I realized it is an issue in Neovim nightly.

### Steps to reproduce

`nvim --clean -u hl.lua`

hl.lua:
```
-- Change icon set on background change
vim.api.nvim_create_autocmd("OptionSet", {
	pattern = "background",
	callback = function()
		vim.api.nvim_set_hl(0, "Normal", {
			fg = "Red",
			ctermfg = "Red",
		})
	end,
})
```

### Expected behavior

Intro screen should be displayed.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-1904+g2498747ad-dirty

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS 14.2 (23C64)

### Terminal name/version

kitty 0.31.0 created by Kovid Goyal

### $TERM environment variable

xterm-kitty

### Installation

build from repo

## Demo

- given simple.vim
```vim
autocmd OptionSet background hi Normal ctermfg=Red guifg=Red
```
- ovi is an alias for vim `v9.0`
- nvim is the nightly version of nvim (`NVIM v0.10.0-dev-1920+gaf93a74a0` at the time of demo)
- bobnvim is an alias for nvim `v0.9.4`

https://github.com/neovim/neovim/assets/10135646/d5dc61a0-0a3a-4088-a289-6c74f060cf2a

**Comment by debdutdeb** [comment]
I could see the same behaviour even without any init files .. `nvim --clean` ..something like `set ft=bash`
```
NVIM v0.10.0-dev-1094d0c0d
Build type: RelWithDebInfo
LuaJIT 2.1.1697887905
Run "nvim -V1 -v" for more info
```

**Comment by debdutdeb** [comment]
Same thing on 0.9.5 . am I misunderstanding the issue?

**Comment by debdutdeb** [comment]
![issue-demo094](https://github.com/neovim/neovim/assets/76006232/1089d23b-10df-4640-9b0c-368f7ace8f47)

**Comment by FelipeIzolan** [comment]
You can reach the same behavior on neovim 0.9:

- Start neovim without opening a file/folder
- :set property=value
- The screen is cleaned.

I don't if it's a bug or not, but neovim 0.10 start setting options in a way that "OptionSet" is triggered.

**Comment by mikesmithgh** [comment]
> You can reach the same behavior on neovim 0.9:
> 
> * Start neovim without opening a file/folder
> * :set property=value
> * The screen is cleaned.
> 
> I don't know if it's a bug or not, but neovim 0.10 start setting options in a way that "OptionSet" is triggered.

If you interact with Neovim, then the intro screen should be cleared. So, typing :set property=value should clear the intro. So I would expect that behavior. You can see the same in vim. I'll add a demo to clarify what exactly I am seeing

**Comment by FelipeIzolan** [comment]
- `"OptionSet"` - is triggered two times at startup (I don't know if it's a bug! on neovim 0.9.4 it not occurs.
- `nvim_set_hl` - is called.
- `redraw_all_later` - all windows is marked to redraw.
- `update_screen` - redraw without maintaining the intro.

**Comment by glepnir** [comment]
can't reproduce please try build with latest commit .your build version is old.

**Comment by FelipeIzolan** [comment]
> can't reproduce please try build with latest commit .your build version is old.

```lua
local c = 0

vim.api.nvim_create_autocmd("OptionSet", {
	pattern = "*",
	callback = function()
          c = c +1
          print(c) 

          vim.api.nvim_set_hl(0, "Normal", {
		fg = "Red",
		ctermfg = "Red",
              })
           
	end,
})
```

https://github.com/neovim/neovim/assets/80170121/1969b834-cccc-4b42-89b5-3ba3d7b04fad

**Comment by glepnir** [comment]
your neovim version ?

**Comment by FelipeIzolan** [comment]
> your neovim version ?

![image](https://github.com/neovim/neovim/assets/80170121/17dafe85-3ad8-476f-ad5e-ada38c5ff358)

**Comment by glepnir** [comment]
thanks for response i can't still reproduce it ...

https://github.com/neovim/neovim/assets/41671631/efbf219a-1e3d-4f59-bb8a-822380fb810b

**Comment by mikesmithgh** [comment]
Hey @glepnir, I still see this on nightly. I tracked this down with a git bisect to 
```
f4f7e294695abe7837af62b68564114de7abd0b4 is the first bad commit
commit f4f7e294695abe7837af62b68564114de7abd0b4
Author: Gregory Anders <8965202+gpanders@users.noreply.github.com>
Date:   Fri Dec 15 16:35:55 2023 -0600

    refactor(defaults): always set options, even if value hasn't changed (#26595)
    
    Comparing against the old value before setting matched the original
    C implementation, but there is no reason to use this restriction. In
    particular, this inhibits using OptionSet to determine when the option
    was set. If users need to handle a case where the option _changed_, it
    is easy to do so in an OptionSet autocommand using v:option_new and
    v:option_old (and friends).

 runtime/lua/vim/_defaults.lua | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)
 ```
 
## demo with intro cleared issue

https://github.com/neovim/neovim/assets/10135646/7ec70a21-66bf-44a6-975e-f3d8e6030ee9

## demo without intro cleared issue

https://github.com/neovim/neovim/assets/10135646/2dde2b62-9056-426e-8a09-384ef041482e


In the case when the intro is not cleared, it doesn't look like the autocommand is triggered.

**Comment by zeertzjq** [comment]
Fixed by #27852
